feat: shifting scene animation for BottomNavigator

andreialecu opened the pull request:
<!-- Please provide enough information so that others can review your pull request. -->
<!-- Keep pull requests small and focused on a single change. -->

### Summary

Adds a nicer optional shifting animation to BottomNavigator than just opacity.

No breaking changes, the old functionality is still preserved. In addition it also allows overriding the Easing function of the animation.

See attached video for a demo.

https://user-images.githubusercontent.com/697707/168426607-29a42d4c-60b5-452e-8cb0-141f96d25894.mov


https://user-images.githubusercontent.com/697707/169238977-14e49a68-e258-433b-80d3-34a787791228.mov


### Test plan

Open the "Bottom Navigation w/ shifting animation" example using the demo app in the repository.

I have verified that `translateY` works properly to remove clipped subviews. You can enable the Perf Monitor and observe the Views counter. You can remove the `translateY` style to test that views won't get clipped, add it back and everything works properly.

andreialecu commented on src/components/BottomNavigation/BottomNavigation.tsx:
```diff
@@ -631,15 +675,23 @@ const BottomNavigation = ({
+                  styles.content,
+                  {
+                    opacity: opacity,
+                    transform: [{ translateX: left }, { translateY: top }],
```
Note that I replaced `{ top }` with `{ translateY }` because `top` cannot be animated by the native driver and it resulted in some errors.

I have verified that `translateY` works properly to remove clipped subviews. You can enable the Perf Monitor and observe the `Views` counter. If you remove `translateY` the views won't be clipped, but with it everything is fine.

lukewalczak commented:
Hey @andreialecu, thanks for your work! Could you please extend the description by adding the video from the Android platform?

lukewalczak commented on src/components/BottomNavigation/BottomNavigation.tsx:
```diff
@@ -203,6 +204,15 @@ type Props = {
+  /**
+   * The scene animation Easing.
+   */
+  sceneAnimationEasing?: EasingFunction | undefined;
```
Please add tests covering new props.

andreialecu commented:
I added the recording and tests. 

However, I noticed that the tests are a bit useless. Because they're written in JavaScript there's no type checking of the props. And most props don't make it to the snapshot.

For example, I changed a random existing test to the following and could still get everything to pass:
<img width="386" alt="Screenshot 2022-05-19 at 10 45 41" src="https://user-images.githubusercontent.com/697707/169240507-ea03299d-f87a-4afc-b825-7fb514432e60.png">


The test I added has the same issue.

lukewalczak commented on src/components/BottomNavigation/BottomNavigation.tsx:
```diff
@@ -442,13 +464,22 @@ const BottomNavigation = ({
-            useNativeDriver: true,
-          })
-        ),
+        ...navigationState.routes.flatMap((_, i) => {
```
Since `flatMap` is a recent API, there can be some issues on older iOS versions. I would suggest confirming whether it's properly supported or rewriting it.

andreialecu replied:
Changed it to two maps instead in https://github.com/callstack/react-native-paper/pull/3176/commits/973a99bbeb2453e5a8da4ac3ff075678e2840b42

lukewalczak commented on src/components/BottomNavigation/BottomNavigation.tsx:
```diff
@@ -447,6 +469,15 @@ const BottomNavigation = ({
             toValue: i === index ? 1 : 0,
             duration: shifting ? 150 * scale : 0,
             useNativeDriver: true,
+            easing: sceneAnimationEasing,
+          })
+        ),
+        ...navigationState.routes.map((_, i) =>
+          Animated.timing(tabsPositionAnims[i], {
+            toValue: i === index ? 0 : i >= index ? 1 : -1,
+            duration: shifting ? 150 * scale : 0,
+            useNativeDriver: true,
+            easing: sceneAnimationEasing,
```
When testing it against animation within Facebook app, I have the feeling animation at the end stops suddenly. I think a more natural behavior should cover decreasing the velocity towards the end before a complete finish (prolly controlled by easing). Could you please to check it and try to refine it?

andreialecu replied:
@lukewalczak maybe something like:

```tsx
      sceneAnimationEasing={Easing.out(Easing.sin)}
      theme={{ animation: { scale: 1.5 } }}
```

It's totally adjustable by the user. The animation is too fast so it needs to be slowed down a bit.

Here's a recording from an actual device with the above settings:


https://user-images.githubusercontent.com/697707/169316516-593327f7-ac6f-4b6a-bcd3-4435e5cd07a2.mov

drplauska commented on example/src/Examples/BottomNavigationAnimatedExample.tsx:
```diff
@@ -0,0 +1,124 @@
+  );
+};
+
+const BottomNavigationExample = () => {
```
`BottomNavigationAnimatedExample`

lukewalczak commented on src/components/__tests__/BottomNavigation.test.js:
```diff
@@ -52,6 +52,24 @@ it('renders shifting bottom navigation', () => {
+    .toJSON();
+
+  expect(tree).toMatchSnapshot();
+});
```
I think you can also add the unit test case, where you're expecting that new props are properly passed and animation is run in the correct sequence.

lukewalczak commented on src/components/BottomNavigation/BottomNavigation.tsx:
```diff
@@ -211,6 +212,15 @@ type Props = {
+   * The scene animation effect. Specify `'shifting'` for a different effect.
+   * By default, 'opacity' will be used.
+   */
+  sceneAnimationType?: 'opacity' | 'shifting';
```
Since those two props will be introduced to v5 can we add a badge annotation: `   * @supported Available in v5.x`. Also, it will be awesome if you can mention them in the [migration guide](https://github.com/callstack/react-native-paper/blob/main/docs/pages/10.migration-guide-to-5.0.md#bottomnavigation-navigation-bar) as new props with their quick functionality description.

andreialecu replied:
done in [callstack/react-native-paper@`e313135` (#3176)](https://github.com/callstack/react-native-paper/pull/3176/commits/e313135dc4f0d69909c8d5859910e4c89b340202)

lukewalczak commented on example/src/Examples/BottomNavigationAnimatedExample.tsx:
```diff
@@ -0,0 +1,140 @@
+import * as React from 'react';
```
We have a convention to have one example file per component, so maybe we can replace the whole file in favour of some kind of control (switch) at the top of `BottomNavigationExample` which toggles new props?

andreialecu replied:
done in [callstack/react-native-paper@`bd35148` (#3176)](https://github.com/callstack/react-native-paper/pull/3176/commits/bd351483d0c8e2616f4f6121ab5af5925c98e0d1)

andreialecu commented on docs/pages/10.migration-guide-to-5.0.md:
```diff
@@ -189,6 +189,12 @@ It's worth to mention that default value of prop `shifting` depends on the theme
+* `sceneAnimationType: "opacity" | "shifting" | undefined` - defines the animation type for the scene. `shifting` enables a new animation where navigating to a scene will shift it horizontally into view. Both `opacity` and `undefined` have the same effect, fading the scene into view.
+* `sceneAnimationEasing` allows specifying a custom easing function for the scene animation.
+
+![shiftingAnimation](screenshots/bottom-navigation-shifting.gif)
```
Not sure if I linked this properly. /cc @lukewalczak

andreialecu replied:
Feel free to edit my PR for any final touches. "Allow edits" is on.
